ddtrace/profiling/collector/_memalloc_reentrant.h

```diff
@@ -1,50 +1,214 @@
+// From the MSVC docs:
+// "_InterlockedCompareExchange64 is available on x86 systems running on any Pentium architecture; it is not
+// available on 386 or 486 architectures."
+static inline uint64_t
+atomic_inc_clamped(uint64_t* target, uint64_t max)
+{
+    // In reality, there's virtually no scenario in which this deadlocks.  Just the same, give it some arbitrarily high
+    // limit in order to prevent unpredicted deadlocks.  96 is chosen since it's the number of cores on the largest
+    // consumer CPU generally used by our customers.
+    int attempts = 96;
+    while (attempts--) {
+        uint64_t old_val = (volatile uint64_t) * target;
+        uint64_t new_val;
 
-#ifdef _PY37_AND_LATER
-extern Py_tss_t memalloc_reentrant_key;
+        // Saturation check
+        if (old_val == UINT64_MAX) {
+            return 0;
+        }
+
+        // CAS loop
+        new_val = old_val + 1;
+#if defined(_MSC_VER)
+        uint64_t prev_val =
+          (uint64_t)InterlockedCompareExchange64((volatile LONG*)target, (LONG)new_val, (LONG)old_val);
+        if (prev_val == old_val) {
+            return new_val;
+        }
+#elif defined(__clang__) || defined(__GNUC__)
+        if (atomic_compare_exchange_strong_explicit(
+              (_Atomic uint64_t*)target, &old_val, new_val, memory_order_seq_cst, memory_order_seq_cst)) {
+            return new_val;
+        }
 #else
-extern int memalloc_reentrant_key;
+#error "Unsupported compiler for atomic operations"
 #endif
+        // If we reach here, CAS failed; another thread changed `target`
+        // Retry until success or until we detect max.
+    }
 
-/* Any non-NULL pointer can be used */
-#define _MEMALLOC_REENTRANT_VALUE Py_True
+    return 0;
+}
 
-static inline void
-memalloc_set_reentrant(bool reentrant)
+// Opaque lock type
+typedef struct
 {
-    if (reentrant)
-#ifdef _PY37_AND_LATER
-        PyThread_tss_set(&memalloc_reentrant_key, _MEMALLOC_REENTRANT_VALUE);
+#ifdef _WIN32
+    HANDLE mutex;
 #else
-        PyThread_set_key_value(memalloc_reentrant_key, _MEMALLOC_REENTRANT_VALUE);
+    pthread_mutex_t mutex;
 #endif
-    else
-#ifdef _PY37_AND_LATER
-        PyThread_tss_set(&memalloc_reentrant_key, NULL);
+} memlock_t;
+
+// Generic initializer
+static inline bool
+memlock_init(memlock_t* lock)
+{
+    if (!lock)
+        return false;
+
+#ifdef _WIN32
+    lock->mutex = CreateMutex(NULL, FALSE, NULL);
+    return lock->mutex != NULL;
+#else
+    return pthread_mutex_init(&lock->mutex, NULL) == 0;
+#endif
+}
+
+// Unlock function
+static inline bool
+memlock_unlock(memlock_t* lock)
+{
+    if (!lock)
+        return false;
+
+#ifdef _WIN32
+    return ReleaseMutex(lock->mutex);
 #else
-        PyThread_set_key_value(memalloc_reentrant_key, NULL);
+    return pthread_mutex_unlock(&lock->mutex) == 0;
 #endif
 }
 
+// Timed lock function
 static inline bool
-memalloc_get_reentrant(void)
+memlock_lock_timed(memlock_t* lock, uint32_t timeout_ms)
```

Comment:
Also, to capture our in-person discussion, your idea was to change this from a timed lock to an un-timed lock and bail immediately if we can't acquire the lock (just in memalloc_add_event? others?) Plus an option (set via env var) to crash if the lock is contended so we can debug why this lock is even needed.

Reply:
Thank you--done.  The environment variable is `_DD_PROFILING_MEMALLOC_CRASH_ON_MUTEX_PASS` and it'll do a case-insensitive check for truthy (1, true, yes, on enable(d))